process_hierarchy.py

```diff
@@ -0,0 +1,478 @@
+    ret_dict = {}
+    for rec in records:
+        ret_dict[rec["id"]] = copy.deepcopy(rec)
+    return ret_dict
+
+
+def create_xdmod_key(high_key, mid_key="", low_key=""):
+    """
+    This combines upto 3 keys into a single (shorter) key
+
+    This was done as a general way of building a key from the
+    datafields and not assiging it a key (which is what I've wound up
+    doing)
+    """
+    m = hashlib.sha256()
+    m.update((f"{str(high_key)}+{str(mid_key)}+{str(low_key)}").encode("utf-8"))
+    return m.hexdigest()
+
+
+def create_hierarchy_files(hierarchy):
+    """
+    This builds the hierarchy files from the data coming from coldfront, keycloak
+    and historic data from the database
+    """
+    # construct hierarchy.csv
+    with open("hierarchy.csv", "w", encoding="utf-8") as hierarchy_file:
+        for rec_id, rec in hierarchy["institution"].items():
+            unique_key = create_xdmod_key(rec_id)
+            hierarchy_file.write(f"{unique_key},{rec['name']},\n")
+        for l2_id, l2 in hierarchy["field-of-science"].items():
+            unique_key = create_xdmod_key(
+                hierarchy["inistitution"][l2["parent_id"]], l2_id
+            )
+            parent_key = create_xdmod_key(l2["parent_id"])
+            hierarchy_file.write(f"{unique_key},{l2['name']},{parent_key}\n")
+        for l3_id, l3_rec in hierarchy["pi"].items():
+            unique_key = create_xdmod_key(
+                (hierarchy["field-of-science"][l3_rec["parent_id"]])["parent_id"],
+                l3_rec["parent_id"],
+                l3_id,
+            )
+            parent_key = create_xdmod_key(
+                (hierarchy["field-of-science"][l3_rec["parent_id"]])["parent_id"],
+                l3_rec["parent_id"],
+            )
+            hierarchy_file.write(f"{unique_key},{l3_rec['name']},{parent_key}\n")
+
+    # constrcut groups.csv (level 4 of the hierarchy, though this is a mapping table between pi/group and the hiearachy)
+    with open("group.csv", "w", encoding="utf-8") as group_file:
+        for l4_id, l4_rec in hierarchy["project"].items():
+            l3_rec = hierarchy["pi"][l4_rec["parent_id"]]
+            unique_key = create_xdmod_key(
+                (hierarchy["field-of-science"][l3_rec["parent_id"]])["parent_id"],
+                l3_rec["parent_id"],
+                l3_rec["id"],
+            )
+            group_file.write(f'"{l4_rec["name"]}", "{unique_key}" \n')
+
+    # constrcut pi2project
+    with open("pi2project.csv", "w", encoding="utf-8") as pi2project_file:
+        for l5 in hierarchy["cloud-project"].values():
+            l4_rec = hierarchy["cloud-project"][l5["parent_id"]]
+            l4_id = l4_rec["id"]
+            pi2project_file.write(
+                f"{hierarchy['cloud-project'][l4_id]['name']}, {l5['name']}\n"
+            )
+
+    # construct the names (rename projects)
+
+
+def find_hierarchy_id(name, dictionary):
```

Comment:
I'm confused by the existence of this function. You already use `pi_id` as the key in the dictionary, therefore you can query PI information just by doing `hierarchy["pi"][pi_id]`. Is there something I'm missing?

Reply:
Only I assign a unique id to the PIs, so the hierarchy["pi"]["id"] is not the same as hierarchy["pi"]["name"].  This makes fetching the PI information a little more complex, however, makes the overall complexity less as I can use the same function to fetch the information at any level in the hierarchy by the name (returning a unique id).  As opposed to doing something special for each layer of the hierarchy.

Additionally, by assigning a unique id, I don't have to do any cryptological digest in order to build a composite key, nor do I have to worry if the field in the xdmod is big enough.